<a href="https://colab.research.google.com/github/cecilia-wanja/Prediction-of-the-people-who-were-likely-to-survive-in-the-Titanic/blob/master/Predicting_spam_emails.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***Question***

Predicting whether or not an email is spam or not.

***Background Information***

Our collection of spam e-mails came from our postmaster and individuals who had filed spam. Our collection of non-spam e-mails came from filed work and personal e-mails, and hence the word 'george' and the area code '650' are indicators of non-spam.

The last column of 'spambase.data' denotes whether the e-mail was considered spam (1) or not (0), i.e. unsolicited commercial e-mail. Most of the attributes indicate whether a particular word or character was frequently occuring in the e-mail. 

In [0]:
#importing the necessary libraries
import pandas as pd

#previewing our dataset
data = pd.read_csv('/content/spambase_csv.csv')
data.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
0            0.00               0.64  ...                       278      1
1            0.21               0.28  ...                      1028      1
2            0.06               0.00  ...                      2259      1
3            0.00               0.00  ...                       191      1
4            0.00               0.00  ...                       191      1

[5 rows x 58 columns]

In [0]:
x= data.iloc[:,:48].values

y=data.iloc[:,-1].values

In [0]:
x.reshape(-1, 1)


array([[0.  ],
       [0.64],
       [0.64],
       ...,
       [0.65],
       [0.  ],
       [0.  ]])

In [0]:
#importing the different naive bayes classifiers
from sklearn.naive_bayes import GaussianNB,MultinomialNB, BernoulliNB

In [0]:
#defining the three models of naive bayes that we are going to use.
gau = GaussianNB()
multi = MultinomialNB()
bern = BernoulliNB()


In [0]:
#dividing our dataset into training and test set

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.2,random_state = 0)

In [0]:
#training using the models above
gau.fit(x_train, y_train)
multi.fit(x_train, y_train)
bern.fit(x_train, y_train)

BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

In [0]:
#predicting the outcome
y_pred1 = gau.predict(x_test)
y_pred2 = multi.predict(x_test)
y_pred3 = bern.predict(x_test)

In [0]:
#evaluating the model

#importing the necessary libraries
from sklearn import metrics

#model accuracy
print('Accuracy of Gaussian:', metrics.accuracy_score(y_test,y_pred1)*100)
print('\n')

print('Accuracy of Multinomial:', metrics.accuracy_score(y_test,y_pred2)*100)
print('\n')

print('Accuracy of Bernoulli:', metrics.accuracy_score(y_test,y_pred3)*100)
print('\n')

#classification report of each
print('Classification report of Gaussian:',)
print(metrics.classification_report(y_test,y_pred1))
print('\n')

print('Classification report of Multinomial:',) 
print(metrics.classification_report(y_test,y_pred2))
print('\n')


print('Classification report of Bernoulli:',)
print(metrics.classification_report(y_test,y_pred3))

Accuracy of Gaussian: 79.58740499457112


Accuracy of Multinomial: 87.07926167209556


Accuracy of Bernoulli: 86.53637350705755


Classification report of Gaussian:
              precision    recall  f1-score   support

           0       0.93      0.70      0.80       538
           1       0.69      0.92      0.79       383

    accuracy                           0.80       921
   macro avg       0.81      0.81      0.80       921
weighted avg       0.83      0.80      0.80       921



Classification report of Multinomial:
              precision    recall  f1-score   support

           0       0.95      0.83      0.88       538
           1       0.79      0.93      0.86       383

    accuracy                           0.87       921
   macro avg       0.87      0.88      0.87       921
weighted avg       0.88      0.87      0.87       921



Classification report of Bernoulli:
              precision    recall  f1-score   support

           0       0.87      0.91      0.89     

**Hyperparameter tuning**

In [0]:
# Multinomial Naive Bayes model parameter tuning
model = MultinomialNB()
param_grid = {'alpha': [1.0, 2.0, 3.0, 4.0, 5.0]}

In [0]:
#Hyper parameter Tuning
from sklearn.model_selection import GridSearchCV

print('Hyper Parameter Tuning')
grid = GridSearchCV(estimator = multi, param_grid = param_grid, cv = 5)

grid.fit(x_train, y_train)

Hyper Parameter Tuning


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=MultinomialNB(alpha=1.0, class_prior=None,
                                     fit_prior=True),
             iid='warn', n_jobs=None,
             param_grid={'alpha': [1.0, 2.0, 3.0, 4.0, 5.0]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [0]:
print("Results returned by GridSearchCV\n")
print("Best estimator: ", grid.best_estimator_)
print("\n")
print("Best score: ", grid.best_score_)
print("\n")
print("Best parameters found: ", grid.best_params_)

Results returned by GridSearchCV

Best estimator:  MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)


Best score:  0.8698369565217391


Best parameters found:  {'alpha': 1.0}


In [0]:
multi = MultinomialNB()


In [0]:
#dividing our dataset into training and test set

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.4,random_state = 0)

In [64]:
multi.fit(x_train, y_train)


MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [0]:
y_pred2 = multi.predict(x_test)


In [66]:
print('Classification report of Multinomial:')
print(metrics.classification_report(y_test,y_pred2))


Classification report of Multinomial:
              precision    recall  f1-score   support

           0       0.95      0.83      0.88      1097
           1       0.79      0.93      0.85       744

    accuracy                           0.87      1841
   macro avg       0.87      0.88      0.87      1841
weighted avg       0.88      0.87      0.87      1841



***Conclusion and Recommendation***

As the test_size increased, there was no trend in the way the accuracy_score changed. Hence the test_size is not an accuracy_score factor.

The best Estimator for this dataset was MultiNomial according to the best estimator function.

On the other hand, hyperparameter tuning did not make a significant difference in the prediction of the model.Instead it reduces it by 1%.

The precision of our 'not spam' dataset was 0.95 which means that the measurement is very precise.The precision of our spam dataset is 0.79 which is not very good. Hence our model will mostly recognize a 'not spam' email as compared to a spam email.